### PR TITLE
Indicate certain string functions cannot fail

### PR DESCRIPTION
**Pull request by martint**
This allows the optimizer to be more aggressive during predicate pushdown.

(x) Release notes are required, with the following suggested text:

```markdown
## General
* Improve performance when using various string functions in queries involving joins. ({issue}`24182`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/ir/IrExpressions.java [hunk 1]
```diff
@@ -119,6 +119,9 @@ private static boolean mayFail(ResolvedFunction function)
                 !name.equals(builtinFunctionName("try_cast")) &&
                 !name.equals(builtinFunctionName("$not")) &&
                 !name.equals(builtinFunctionName("substring")) &&
+                !name.equals(builtinFunctionName("trim")) &&
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/ir/IrExpressions.java:122]: There's other string functions like `lpad` (and functions other than string functions), what's the criteria for something getting on this list and is there a way to check if we're missing more entries here ?

Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/ir/IrExpressions.java:122]: Functions that are guaranteed not to fail for any input should be in this list. 

Eventually, we need to move this to be an attribute of the functions that the optimizer can inspect. So this is a stopgap.

Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/sql/ir/IrExpressions.java:122]: Can we add other commonly used string functions like upper, lower, lpad, rpad, reverse etc. too ?
Does it make sense to extend `allow_unsafe_pushdown` to allow bypass this check as well ?

**Review by raunaqmorarka — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/sql/ir/IrExpressions.java (hunk 1 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/ir/IrExpressions.java:122]: Yeah, I'll add other functions.

> Does it make sense to extend allow_unsafe_pushdown to allow bypass this check as well?

No. That check is a temporary workaround for something that has no solution in the current optimizer. For functions, I'd rather add them here.

**Review by kasiafi — APPROVED** [review]
Noted the request for "never_fail" attribute.

**Pull request merged**